### PR TITLE
Revert the reverts

### PR DESCRIPTION
**Pull request by malfet**
Revert all previous reverts but "Make tls_local_dispatch_key_set inlineable", which is likely the culprit for Windows failures

### CODE REVIEW

File: torch/csrc/autograd/engine.h [hunk 1]
```diff
@@ -37,7 +37,7 @@ static constexpr int CPU_DEVICE = -1;
 // As we hold mutex in every of our custom C++ autograd Node, we would
 // like to avoid TSAN complains on this when doing reentrant backwards
 // For reference, see https://github.com/google/sanitizers/issues/950
-static constexpr int MAX_DEPTH = 60;
+static constexpr int MAX_DEPTH = 40;
```
Comment (albanD) [inline torch/csrc/autograd/engine.h:40]: @swolchok why was this changed?

File: aten/src/ATen/native/TensorShape.cpp [hunk 2]
```diff
@@ -121,19 +124,21 @@ Tensor & _cat_out_cpu(Tensor& result, TensorList tensors, int64_t dim) {
   }
   at::assert_no_internal_overlap(result);
 
-  auto should_skip = [](const Tensor& t) { return t.numel() == 0 && t.dim() == 1; };
-  for (auto const &tensor : tensors) {
-    if (should_skip(tensor)) {
-      continue;
+  const Tensor* pnotSkippedTensor = [](const TensorList& tensors) -> const Tensor* {
```
Comment (albanD) [inline aten/src/ATen/native/TensorShape.cpp:127]: @swolchok It would be nice here to have a comment that explains why it is ok to take a plain pointer here and why we are sure all the Tensors are kept alive until the end of this function.

File: torch/csrc/autograd/engine.h (hunk 1 above)
Comment (malfet) [reply torch/csrc/autograd/engine.h:40]: @albanD this is my change, see https://github.com/pytorch/pytorch/issues/49558#issuecomment-747840440
For some reason after those inlining changes we started to hit  this one on Windows (which have default stack size of 1Mb) but never on Linux/Mac (with has a larger default stack size, at least 2Mb)

File: aten/src/ATen/native/TensorShape.cpp (hunk 2 above)
Comment (malfet) [reply aten/src/ATen/native/TensorShape.cpp:127]: Again, this is my change, but I agree with you, a comment will be nice.

Comment (swolchok) [reply aten/src/ATen/native/TensorShape.cpp:127]: 1) there's no reason to pass TensorList by reference; it's a reference type already
2) If a reference is OK, a pointer is OK

File: torch/csrc/autograd/engine.h (hunk 1 above)
Comment (swolchok) [reply torch/csrc/autograd/engine.h:40]: please revert this, and pull the CoinflipTLS change since that's likely the one causing the problems

**Review by swolchok — CHANGES_REQUESTED** [review]
please pull the CoinflipTLS change

File: aten/src/ATen/native/TensorShape.cpp (hunk 2 above)
Comment (albanD) [reply aten/src/ATen/native/TensorShape.cpp:127]: > If a reference is OK, a pointer is OK

I am sure it is ok, I am just saying that we should usually have a good reason before using plain pointers. So having a comment here will help future readers understand easily why we do it here.

**Comment by malfet** [comment]
@swolchok I've removed that revert from commit stack, let's see if it passes Windows CI without it

**Comment by peterjc123** [comment]
@malfet You'll need to cherry-pick the fix https://github.com/pytorch/pytorch/pull/49615 here.
